### DISCUSSION TITLE
pyright server insists on sending type 2 completion (snippet) even if the client is replying that it doesn't support "snippets"

### DISCUSSION THREAD

**Discussion by tshu-w** [microsoft/pyright#3197] [Q&A]
Hi, pyright server seems insist on sending type 2 completion (snippet) even if the client is replying that it doesn't support "snippets" as analysis in https://github.com/joaotavora/eglot/issues/884#issuecomment-1067698924

**Comment by erictraut** [comment]
I'm not sure what you mean by "snippets" in this context. Can you provide more context?

Which LSP client are you using? And what is the LSP feature it claims not to support?

**Reply by tshu-w** [reply]
I'm using eglot on Emacs with pyright. If I type the following code in `test.py`:
```python
class Demo:
    def __i
```
pyright would try to reply `kind 2 completionItem`(see `completionItemKind` in https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion ):
```
[client-request] (id:19) Tue Mar 15 09:15:53 2022:
(:jsonrpc "2.0" :id 19 :method "textDocument/completion" :params
	  (:textDocument
	   (:uri "file:///Users/wangtianshu/dotfiles/test.py")
	   :position
	   (:line 1 :character 11)
	   :context
	   (:triggerKind 1)))
[server-reply] (id:19) Tue Mar 15 09:15:53 2022:
(:jsonrpc "2.0" :id 19 :result
	  (:items
	   [(:label "__init__" :kind 2 :data
		    (:workspacePath "/Users/wangtianshu/dotfiles/" :filePath "/Users/wangtianshu/dotfiles/test.py" :position
				    (:line 1 :character 11)
				    :funcParensDisabled t :symbolLabel "__init__")
		    :sortText "11.9999.__init__" :textEdit
		    (:range
		     (:start
		      (:line 1 :character 8)
		      :end
		      (:line 1 :character 11))
		     :newText "__init__(self) -> None:\n    pass"))
	    (:label "__getattribute__" :kind 2 :data
		    (:workspacePath "/Users/wangtianshu/dotfiles/" :filePath "/Users/wangtianshu/dotfiles/test.py" :position
				    (:line 1 :character 11)
				    :funcParensDisabled t :symbolLabel "__getattribute__")
		    :sortText "11.9999.__getattribute__" :textEdit
		    (:range
		     (:start
		      (:line 1 :character 8)
		      :end
		      (:line 1 :character 11))
		     :newText "__getattribute__(self, __name: str) -> Any:\n    pass"))
	    (:label "__sizeof__" :kind 2 :data
		    (:workspacePath "/Users/wangtianshu/dotfiles/" :filePath "/Users/wangtianshu/dotfiles/test.py" :position
				    (:line 1 :character 11)
				    :funcParensDisabled t :symbolLabel "__sizeof__")
		    :sortText "11.9999.__sizeof__" :textEdit
		    (:range
		     (:start
		      (:line 1 :character 8)
		      :end
		      (:line 1 :character 11))
		     :newText "__sizeof__(self) -> int:\n    pass"))
	    (:label "__dir__" :kind 2 :data
		    (:workspacePath "/Users/wangtianshu/dotfiles/" :filePath "/Users/wangtianshu/dotfiles/test.py" :position
				    (:line 1 :character 11)
				    :funcParensDisabled t :symbolLabel "__dir__")
		    :sortText "11.9999.__dir__" :textEdit
		    (:range
		     (:start
		      (:line 1 :character 8)
		      :end
		      (:line 1 :character 11))
		     :newText "__dir__(self) -> Iterable[str]:\n    pass"))
	    (:label "__init_subclass__" :kind 2 :data
		    (:workspacePath "/Users/wangtianshu/dotfiles/" :filePath "/Users/wangtianshu/dotfiles/test.py" :position
				    (:line 1 :character 11)
				    :funcParensDisabled t :symbolLabel "__init_subclass__")
		    :sortText "11.9999.__init_subclass__" :textEdit
		    (:range
		     (:start
		      (:line 1 :character 8)
		      :end
		      (:line 1 :character 11))
		     :newText "__init_subclass__(cls) -> None:\n    pass"))
	    (:label "__instancecheck__" :kind 2 :data
		    (:workspacePath "/Users/wangtianshu/dotfiles/" :filePath "/Users/wangtianshu/dotfiles/test.py" :position
				    (:line 1 :character 11)
				    :funcParensDisabled t :symbolLabel "__instancecheck__")
		    :sortText "11.9999.__instancecheck__" :textEdit
		    (:range
		     (:start
		      (:line 1 :character 8)
		      :end
		      (:line 1 :character 11))
		     :newText "__instancecheck__(self, __instance: Any) -> bool:\n    pass"))]
	   :isIncomplete t))
```
However, the client have already say it doesn't suppor through the `:completion :snippetSupport` section:
```
[client-request] (id:1) Tue Mar 15 09:15:42 2022:
(:jsonrpc "2.0" :id 1 :method "initialize" :params
	  (:processId 81846 :rootPath "/Users/wangtianshu/dotfiles/" :rootUri "file:///Users/wangtianshu/dotfiles" :initializationOptions #s(hash-table size 65 test eql rehash-size 1.5 rehash-threshold 0.8125 data
																			())
		      :capabilities
		      (:workspace
		       (:applyEdit t :executeCommand
				   (:dynamicRegistration :json-false)
				   :workspaceEdit
				   (:documentChanges t)
				   :didChangeWatchedFiles
				   (:dynamicRegistration t)
				   :symbol
				   (:dynamicRegistration :json-false)
				   :configuration t)
		       :textDocument
		       (:synchronization
			(:dynamicRegistration :json-false :willSave t :willSaveWaitUntil t :didSave t)
			:completion
			(:dynamicRegistration :json-false :completionItem
					      (:snippetSupport :json-false :deprecatedSupport t :tagSupport
							       (:valueSet
								[1]))
					      :contextSupport t)
			:hover
			(:dynamicRegistration :json-false :contentFormat
					      ["plaintext"])
			:signatureHelp
			(:dynamicRegistration :json-false :signatureInformation
					      (:parameterInformation
					       (:labelOffsetSupport t)
					       :activeParameterSupport t))
			:references
			(:dynamicRegistration :json-false)
			:definition
			(:dynamicRegistration :json-false :linkSupport t)
			:declaration
			(:dynamicRegistration :json-false :linkSupport t)
			:implementation
			(:dynamicRegistration :json-false :linkSupport t)
			:typeDefinition
			(:dynamicRegistration :json-false :linkSupport t)
			:documentSymbol
			(:dynamicRegistration :json-false :hierarchicalDocumentSymbolSupport t :symbolKind
					      (:valueSet
					       [1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26]))
			:documentHighlight
			(:dynamicRegistration :json-false)
			:codeAction
			(:dynamicRegistration :json-false :codeActionLiteralSupport
					      (:codeActionKind
					       (:valueSet
						["quickfix" "refactor" "refactor.extract" "refactor.inline" "refactor.rewrite" "source" "source.organizeImports"]))
					      :isPreferredSupport t)
			:formatting
			(:dynamicRegistration :json-false)
			:rangeFormatting
			(:dynamicRegistration :json-false)
			:rename
			(:dynamicRegistration :json-false)
			:publishDiagnostics
			(:relatedInformation :json-false :codeDescriptionSupport :json-false :tagSupport
					     (:valueSet
					      [1 2])))
		       :experimental #s(hash-table size 65 test eql rehash-size 1.5 rehash-threshold 0.8125 data
						   ()))))
```

**Reply by erictraut** [reply]
I think you might be confusing the CompletionItemKind with the InsertTextFormat. The CompletionItemKind = 2 indicates that the completion item is a "method". The InsertTextFormat = 2 indicates that it's a snippet format. I don't see any evidence that pyright is returning a snippet format. I've reviewed the code, and it is honoring the client capabilities flag passed as "capabilities.textDocument.completion.completionItem.snippetSupport".

**Reply by tshu-w** [reply]
Sorry, I'm not familar with lsp. 

@joaotavora Hi, can you take a look at @erictraut reply?

**Reply by joaotavora** [reply]
Hello. I've just become aware of this mention. I've reviewed @erictraut's assessment and he his right. CompletionItemKind = 2 indicates that the completion item is a "method".  Sorry for the confusion.

**Reply by tshu-w** [reply]
@erictraut Hi I would like to know if for lsp server this time should return the correct indent when returning multi-line completions. @joaotavora thinks indent should be returned, but pyright does not return the indent of the previous level (i.e. only four spaces instead of eight in this case).

**Reply by erictraut** [reply]
Can you provide a concrete example? If I understand your question correctly, I would expect the language server to use the appropriate indentation. For example, pyright attempts to determine the most common indentation strategy within a source file (tabs vs. a specific number of spaces) and use that when inserting text for completions. It's possible that we have a big in a specific completion path.

**Reply by tshu-w** [reply]
HI, here is a demo:
```python
class Demo:
    def __i
```
when I typing this, pyright return items (you can find the whole server reply in my above comment):
```
	   [(:label "__init__" :kind 2 :data
		    (:workspacePath "/Users/wangtianshu/dotfiles/" :filePath "/Users/wangtianshu/dotfiles/test.py" :position
				    (:line 1 :character 11)
				    :funcParensDisabled t :symbolLabel "__init__")
		    :sortText "11.9999.__init__" :textEdit
		    (:range
		     (:start
		      (:line 1 :character 8)
		      :end
		      (:line 1 :character 11))
		     :newText "__init__(self) -> None:\n    pass"))
```
there are only 4 spaces before the `pass` so I will get:
```python
class Demo:
    def __init__(self):
    pass
```
eglot hope to have 8 to make it indent correctly. I wander should this be done on server or client? Or it's undefined.

**Reply by erictraut** [reply]
Ah, interesting question. In the case where the client supports snippets, the client should handle this. In the case where the client doesn't handle snippets, I'm not sure but my guess is that the server should handle it. Pyright doesn't currently do so. Could you please file a bug report on this? I can work on a fix.

Do you plan to add support for snippets? That would be preferable since most other clients already provide such support.

**Reply by tshu-w** [reply]
I'm using snippets packages to make sure it indents correctly on Emacs. But I hope this can work without snippets, and It seems pyright doesn't support snippet yet? I will report a bug later this day.

**Reply by erictraut** [reply]
Pyright's completion provide does emit snippets if the client supports them. See https://github.com/microsoft/pyright/blob/043c5bdfca941e773682700541f0b4a8bb9c4b5f/packages/pyright-internal/src/languageService/completionProvider.ts#L1235.

**Reply by erictraut** [reply]
I dug into this further, and VS Code appears to insert the indentation/spaces automatically even if the result is not a snippet. This implies that the client should treat `\n` characters as if the user has typed a "return" at that point and insert the appropriate indentation.

I'm going to assume this is the intended/expected behavior for clients, in which case pyright is correct in not providing additional spaces in its response. So there's no need to file a bug. You'll need to address this in the client code.

**Reply by tshu-w** [reply]
Ok, thank you for your investigating.

**Comment by joaotavora** [comment]
> This implies that the client should treat \n characters as if the user has typed a "return" at that point and insert the appropriate indentation.

I'm sorry to disagree, but I don't know how VS Code's behaviour "implies" it.  If you find something in the LSP spec that is one thing, but VS Code's behaviour is not the authority here, as much as I understand it is Microsoft's editor and you work for/collaborate with Microsoft.

According to my understanding of LSP, an edit is emitted by the server for a given document to leave that document at a certain intended state.  A language agnostic client cannot generally know what the "appropriate indentation" after each newline character provided in the edit.  It's the server who has all knowledge of what the language is or what that indentation means.  In Python, it is not just decorative, as we know.

Snippets are special edits, which are specified to be inserted at a given column and inherit "dumb" current columnar indentation.  At least Emacs's snippets are. They were inherited from the old TextMate editor, as were LSP's.  And in TextMate, they behaved like that.

Anyway, with snippets all is fine: server and client agree on the exact way to do things.  But normal edits are different, they have no such precise indication of behavior. 

At least not until relatively recently.  I went to check the standard, as I should have done in the beginning.  From 3.16.0 on, completion items may contain this nugget:

```ts
/**
	 * How whitespace and indentation is handled during completion
	 * item insertion. If not provided the client's default value depends on
	 * the `textDocument.completion.insertTextMode` client capability.
	 *
	 * @since 3.16.0
	 * @since 3.17.0 - support for `textDocument.completion.insertTextMode`
	 */
	insertTextMode?: InsertTextMode;
```

In https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#insertTextMode you find out more. The server can supply a completion for which "dumb" columnar indentation is to be added by the client:

```ts
/**
 * How whitespace and indentation is handled during completion
 * item insertion.
 *
 * @since 3.16.0
 */
export namespace [InsertTextMode](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#insertTextMode) {
	/**
	 * The insertion or replace strings is taken as it is. If the
	 * value is multi line the lines below the cursor will be
	 * inserted using the indentation defined in the string value.
	 * The client will not apply any kind of adjustments to the
	 * string.
	 */
	export const asIs: 1 = 1;

	/**
	 * The editor adjusts leading whitespace of new lines so that
	 * they match the indentation up to the cursor of the line for
	 * which the item is accepted.
	 *
	 * Consider a line like this: <2tabs><cursor><3tabs>foo. Accepting a
	 * multi line completion item is indented using 2 tabs and all
	 * following lines inserted will be indented using 2 tabs as well.
	 */
	export const adjustIndentation: 2 = 2;
}
```

Of course, as always, the server has to check for the client's capability to handle this new feature.  Probably VS code declares it, but Eglot doesn't.  So the server shouldn't rely on its presence. 

In its absence it *should*, IMHO,  fallback to the old "asIs" indentation semantics. Why, given the legal void, shouldn't it fallback to the other "dumb column" semantics?  Well, because it flows from the even older definition of `textEdit`, which describes the process to insert a given set of characters in a given order in the document.

In practical terms, it may be that Eglot develops this capability in the future.  It doesn't seem devilish to implement and I will support any good PR doing it. But I'm also not particularly concerned:

* this behaviour that I personally consider incorrect seems only to be exhibited by pyright.
* with snippets everything's correctly specified and working, so for now a workaround is just to enable snippets in Emacs, which is easy.

**Reply by tshu-w** [reply]
@erictraut Hi, what's your opinion about @joaotavora comment.

**Reply by erictraut** [reply]
I don't have plans to change the non-snippets behavior at this time. If the emacs LSP plans to enable snippets support, that sounds like it addresses everyone's needs.

**Reply by joaotavora** [reply]
To be clear, Emacs LSP client already supports snippets.  But they can be turned off by the user, which is what I suppose @tshu-w does.
